Begin using C++ coroutines for WebFullScreenManagerProxy IPC functions

achristensen07 opened the pull request:
#### afd0493af45e56cf1005c046ca57af404a96d59d
<pre>
Begin using C++ coroutines for WebFullScreenManagerProxy IPC functions
<a href="https://bugs.webkit.org/show_bug.cgi?id=288413">https://bugs.webkit.org/show_bug.cgi?id=288413</a>
<a href="https://rdar.apple.com/145511185">rdar://145511185</a>

Reviewed by Abrar Rahman Protyasha.

I tried this in 278995@main but it got reverted because it leaked a call frame every time we called a coroutine.
I fixed that by making Task::final_suspend return std::suspend_never because when we call a coroutine
and give it a CompletionHandler, the caller has no more business suspending it.  The coroutine should continue
until it calls the CompletionHandler.  My technique of using a class CoroutineCaller to try to destroy the call
frame from the caller is also not needed.  You just call a coroutine that returns a Task like you do any function,
and it takes care of itself and returns at the first co_await or co_return or the end, whichever comes first.

I made Lazy and Awaitable [[nodiscard]] to prevent bugs where the result isn&apos;t co_awaited.

This makes adding async tasks simpler.  Instead of indenting everything another layer and explicitly copying all
the parameters needed into yet another lambda capture, we just add a co_await doSomethingAsync() call and it just works.
It also makes it so the code won&apos;t compile if we forget to co_return a value, instead of the status quo where
we can forget to call a CompletionHandler in a branch.

WebPage::createAppHighlightInSelectedRange had an unused return value, and I changed it to void because I now
use sfinae to detect if the return value is void when determining which version of callMemberFunction to call.

* Source/WebKit/Platform/CoroutineUtilities.h:
(WebKit::Task::promise_type::get_return_object):
(WebKit::Lazy::PromiseBase::initial_suspend): Deleted.
(WebKit::Lazy::PromiseBase::unhandled_exception): Deleted.
(WebKit::Lazy::PromiseBase::setHandle): Deleted.
(WebKit::Lazy::PromiseBase::handle): Deleted.
(WebKit::Lazy::Promise&lt;U&gt;::get_return_object): Deleted.
(WebKit::Lazy::Promise&lt;U&gt;::return_void): Deleted.
(WebKit::Lazy::Promise&lt;U&gt;::result): Deleted.
(WebKit::Lazy::Awaitable::Awaitable): Deleted.
(WebKit::Lazy::Awaitable::await_ready const): Deleted.
(WebKit::Lazy::Awaitable::await_suspend): Deleted.
(WebKit::Lazy::Awaitable::await_resume): Deleted.
(WebKit::Lazy::Lazy): Deleted.
(WebKit::Lazy::operator co_await const): Deleted.
(WebKit::CoroutineCaller::setCoroutine): Deleted.
(WebKit::callCoroutine): Deleted.
(WebKit::AwaitableTaskWithCompletionHandler::AwaitableTaskWithCompletionHandler): Deleted.
(WebKit::AwaitableTaskWithCompletionHandler::await_ready): Deleted.
(WebKit::AwaitableTaskWithCompletionHandler::await_suspend): Deleted.
(WebKit::AwaitableTaskWithCompletionHandler::await_resume): Deleted.
* Source/WebKit/Platform/IPC/HandleMessage.h:
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.cpp:
(WebKit::WebFullScreenManagerProxy::prepareQuickLookImageURL const):
(WebKit::WebFullScreenManagerProxy::beganEnterFullScreen):
(WebKit::WebFullScreenManagerProxy::beganExitFullScreen):
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::nextPresentationUpdate):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Tools/TestWebKitAPI/CoroutineUtilities.h:

Canonical link: <a href="https://commits.webkit.org/291846@main">https://commits.webkit.org/291846@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3f991f1953b5e8b255a2b0879b9c36ab979e41ee

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/92211 "11 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/11741 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/1616 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/97233 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/42754 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/94261 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/12055 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/20224 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/70728 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28184 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/95212 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/9172 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/83745 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51058 "Found 1 new API test failure: /WPE/TestWebKitWebView:/webkit/WebKitWebView/save (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8891 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/1382 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/42087 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/84956 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/79212 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/1323 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/99274 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/90907 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/19295 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/14258 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/79740 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/19547 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/79625 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78998 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/23526 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/965 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/12299 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15107 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/19276 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/24448 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/113530 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18968 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/32846 "Found 309 new JSC stress test failures: jsc-layout-tests.yaml/js/script-tests/function-apply-many-args.js.layout-no-llint, microbenchmarks/array-from-object.js.lockdown, microbenchmarks/array-prototype-with-storage.js.bytecode-cache, microbenchmarks/array-prototype-with-storage.js.default, microbenchmarks/array-prototype-with-storage.js.dfg-eager, microbenchmarks/array-prototype-with-storage.js.dfg-eager-no-cjit-validate, microbenchmarks/array-prototype-with-storage.js.eager-jettison-no-cjit, microbenchmarks/array-prototype-with-storage.js.lockdown, microbenchmarks/array-prototype-with-storage.js.mini-mode, microbenchmarks/array-prototype-with-storage.js.no-cjit-collect-continuously ... (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/22426 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/20711 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/UIProcess/WebFullScreenManagerProxy.cpp:
```diff
@@ -211,19 +212,26 @@ void WebFullScreenManagerProxy::enterFullScreen(IPC::Connection& connection, Fra
+    Awaitable<bool>::Callback callback { [=] (auto completionHandler) {
+        client->enterFullScreen(mediaDetails.mediaDimensions, WTFMove(completionHandler));
+    } };
+    bool success = co_await Awaitable<bool> { WTFMove(callback) };
```
Do you need to instantiate the callback separately?

aprotyas replied:
Nevermind, we just talked about a clang vs gcc discrepancy.

aprotyas commented on Source/WebKit/Platform/CoroutineUtilities.h:
```diff
@@ -131,55 +120,48 @@ class Lazy {
 
-class CoroutineCaller {
-    WTF_MAKE_TZONE_ALLOCATED_INLINE(CoroutineCaller);
+template<typename> class Awaitable;
```
Is this forward decl necessary? You're just defining the whole class right after anyways, no?

aprotyas commented on Source/WebKit/Platform/CoroutineUtilities.h:
```diff
@@ -131,55 +120,48 @@ class Lazy {
+    using Callback = CompletionHandler<void(CompletionHandler<void(T&&)>)>;
+    Awaitable(Callback&& callback)
+        : m_callback(WTFMove(callback)) { }
+    bool await_ready() { return !!m_result; }
```
`std::optional<T>` has an `operator bool()` so I think it's safe to:

```
bool await_ready() const { return m_result; }
```

aprotyas replied:
also, make it `const`?

achristensen07 replied:
const yes, !! needs to stay.  std::optional::operator bool is explicit.

aprotyas reviewed (approved):
Mainly one comment inline, and another general style suggestion: Can you remove the extra space between the capture list and the argument list in the lambdas? To align with our style guide

achristensen07 committed:
Begin using C++ coroutines for WebFullScreenManagerProxy IPC functions
https://bugs.webkit.org/show_bug.cgi?id=288413
rdar://145511185

Reviewed by Abrar Rahman Protyasha.

I tried this in 278995@main but it got reverted because it leaked a call frame every time we called a coroutine.
I fixed that by making Task::final_suspend return std::suspend_never because when we call a coroutine
and give it a CompletionHandler, the caller has no more business suspending it.  The coroutine should continue
until it calls the CompletionHandler.  My technique of using a class CoroutineCaller to try to destroy the call
frame from the caller is also not needed.  You just call a coroutine that returns a Task like you do any function,
and it takes care of itself and returns at the first co_await or co_return or the end, whichever comes first.

I made Lazy and Awaitable [[nodiscard]] to prevent bugs where the result isn't co_awaited.

This makes adding async tasks simpler.  Instead of indenting everything another layer and explicitly copying all
the parameters needed into yet another lambda capture, we just add a co_await doSomethingAsync() call and it just works.
It also makes it so the code won't compile if we forget to co_return a value, instead of the status quo where
we can forget to call a CompletionHandler in a branch.

WebPage::createAppHighlightInSelectedRange had an unused return value, and I changed it to void because I now
use sfinae to detect if the return value is void when determining which version of callMemberFunction to call.

* Source/WebKit/Platform/CoroutineUtilities.h:
(WebKit::Task::promise_type::get_return_object):
(WebKit::Lazy::PromiseBase::initial_suspend): Deleted.
(WebKit::Lazy::PromiseBase::unhandled_exception): Deleted.
(WebKit::Lazy::PromiseBase::setHandle): Deleted.
(WebKit::Lazy::PromiseBase::handle): Deleted.
(WebKit::Lazy::Promise<U>::get_return_object): Deleted.
(WebKit::Lazy::Promise<U>::return_void): Deleted.
(WebKit::Lazy::Promise<U>::result): Deleted.
(WebKit::Lazy::Awaitable::Awaitable): Deleted.
(WebKit::Lazy::Awaitable::await_ready const): Deleted.
(WebKit::Lazy::Awaitable::await_suspend): Deleted.
(WebKit::Lazy::Awaitable::await_resume): Deleted.
(WebKit::Lazy::Lazy): Deleted.
(WebKit::Lazy::operator co_await const): Deleted.
(WebKit::CoroutineCaller::setCoroutine): Deleted.
(WebKit::callCoroutine): Deleted.
(WebKit::AwaitableTaskWithCompletionHandler::AwaitableTaskWithCompletionHandler): Deleted.
(WebKit::AwaitableTaskWithCompletionHandler::await_ready): Deleted.
(WebKit::AwaitableTaskWithCompletionHandler::await_suspend): Deleted.
(WebKit::AwaitableTaskWithCompletionHandler::await_resume): Deleted.
* Source/WebKit/Platform/IPC/HandleMessage.h:
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.cpp:
(WebKit::WebFullScreenManagerProxy::prepareQuickLookImageURL const):
(WebKit::WebFullScreenManagerProxy::beganEnterFullScreen):
(WebKit::WebFullScreenManagerProxy::beganExitFullScreen):
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::nextPresentationUpdate):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Tools/TestWebKitAPI/CoroutineUtilities.h:

Canonical link: https://commits.webkit.org/291846@main
